### PR TITLE
[IMP] sale_stock: avoid computing expected dates twice

### PR DESCRIPTION
**Pull request by grindtildeath**
Computation of sale.order.expected_date is done in sale module and overridden in sale_stock.
Both function are doing mostly the same, except when it is selecting the min or max date from the sale order lines dates list based on the picking policy.

Use a dedicated function to chose the final date out of the SOL dates, to simplify the override (and avoid calling the method `_expected_date` twice for each line.

Also use `_expected_date` when preparing the procurement, to avoid duplicated code (and ease overrides).

### CODE REVIEW

**Comment by grindtildeath** [comment]
ping @jem-odoo @nim-odoo @Feyensv 

Can you one of you guys please have a look? :pray:

File: addons/sale/models/sale_order.py [hunk 1]
```diff
@@ -556,13 +556,18 @@ def _compute_expected_date(self):
                 order.expected_date = False
                 continue
             dates_list = order.order_line.filtered(
-                lambda line: not line.display_type and not line._is_delivery()
+                lambda line: not line.display_type and not line._is_delivery() and line.state != 'cancel'
             ).mapped(lambda line: line and line._expected_date())
             if dates_list:
-                order.expected_date = min(dates_list)
+                order.expected_date = order._expected_date_list_select_func()(dates_list)
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:562]: I don't really like handling function/methods like that.
Can't you add a method `_select_expected_date(self, expected_dates)` instead, that would be overridden in `sale_stock` to select the `min` or the `max` depending on the `picking_policy` ?

File: addons/sale_stock/models/sale_order.py [hunk 2]
```diff
@@ -87,14 +87,10 @@ def _compute_delivery_status(self):
...
-            dates_list = []
-            for line in order.order_line.filtered(lambda x: x.state != 'cancel' and not x._is_delivery() and not x.display_type):
-                dt = line._expected_date()
-                dates_list.append(dt)
-            if dates_list:
-                expected_date = min(dates_list) if order.picking_policy == 'direct' else max(dates_list)
-                order.expected_date = fields.Datetime.to_string(expected_date)
+
+    def _expected_date_list_select_func(self):
+        super(SaleOrder, self)._expected_date_list_select_func()
```
Comment (Feyensv) [inline addons/sale_stock/models/sale_order.py:92]:
```suggestion
        super()._expected_date_list_select_func()
```

**Review by Feyensv — CHANGES_REQUESTED** [review]
Change seems legit, but is there a real (performance) issue here ?
If not, I believe this should target `master` instead.

**Comment by grindtildeath** [comment]
@Feyensv  Thank you very much for the review. :hugs: 

I amended the commit according to your suggestions.

There is no real performance issue with the standard implementation. However, we did customize `sale.order.line._expected_date` for a customer based on calendars and the computation is getting quite heavy. That's why a single call to `_expected_date` on this compute already helps a lot.

Anyway, if you want to merge this only in master, I'd prefer to open another PR for master. That allows us to keep targeting 16.0 here, to integrate it as a patch in our project.

Please let me know if I should close this one to open a new one targeting master.

**Comment by grindtildeath** [comment]
@Feyensv I realized I had to add an extra commit to use `_expected_date` in the preparation of procurement value as having duplicated code there was breaking our customization.

IMO it makes sense to fallback on this function instead of repeating the same logic.

File: addons/sale_stock/models/sale_order.py [hunk 3]
```diff
@@ -87,14 +87,10 @@ def _compute_delivery_status(self):
...
-            for line in order.order_line.filtered(lambda x: x.state != 'cancel' and not x._is_delivery() and not x.display_type):
-                dt = line._expected_date()
-                dates_list.append(dt)
-            if dates_list:
-                expected_date = min(dates_list) if order.picking_policy == 'direct' else max(dates_list)
-                order.expected_date = fields.Datetime.to_string(expected_date)
+
+    def _select_expected_date(self, expected_dates):
+        res = super()._select_expected_date()
+        return res if self.picking_policy == 'direct' else max(expected_dates)
```
Comment (Feyensv) [inline addons/sale_stock/models/sale_order.py:93]:
```suggestion
        if self.picking_policy == 'direct:
            return super()._select_expected_date(expected_dates)
        return max(expected_dates)
```

You forgot to pass the arg to the super call.
Also, let's avoid computing the `min` if we know we will return the `max`.

**Pull request closed without merging**